### PR TITLE
Vectorize predicates over full col filter scans

### LINKED ISSUE #8634: Consider doing bulk processing in SV scan iterator
```
`
public int next() {
  while (_nextDocId < _numDocs) {
    int nextDocId = _nextDocId++;
    _numEntriesScanned++;
    if (_valueMatcher.doesValueMatch(nextDocId)) {
      return nextDocId;
    }
  }
  return Constants.EOF;
}
`
```

https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/SVScanDocIdIterator.java

Bulk API can potentially be added to do this in a tight-loop (vectorized like manner) for few fixed number of records at a time as opposed to a function call to `doesValueMatch` per document. May be there is way to change interface such that caller of iterator can possibly directly `doesValueMatch` in a tight-loop

### PR DESCRIPTION
**Pull request by richardstartin**
Vectorizes predicate evaluation over full column scans

Fixes #8634

### CODE REVIEW

**Comment by richardstartin** [comment]
Converting to draft because this is a bit slower than before, more attention needs to be paid to cases where the docId is advanced (it may be necessary to vectorize the iteration all the way up to the top)

**Comment by richardstartin** [comment]
This leads to a modest improvement, which would improve when the filter can't be inlined, and if the filter were implemented in a branch-free style

```
Benchmark               (_numRows)                                                                 (_query)  (_scenario)  Mode  Cnt   Score   Error  Units
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.001)  avgt    5  13.447 ± 1.817  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10     EXP(0.5)  avgt    5  30.589 ± 0.710  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.999)  avgt    5  25.331 ± 2.261  ms/op

Benchmark               (_numRows)                                                                 (_query)  (_scenario)  Mode  Cnt   Score   Error  Units
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.001)  avgt    5  16.033 ± 5.131  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10     EXP(0.5)  avgt    5  36.588 ± 1.894  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.999)  avgt    5  31.201 ± 3.927  ms/op
```

**Comment by richardstartin** [comment]
Adding a couple of queries
```java
    getBrokerResponse("SELECT RAW_INT_COL FROM MyTable "
        + "WHERE RAW_INT_COL = 10");
    getBrokerResponse("SELECT RAW_INT_COL FROM MyTable "
        + "WHERE RAW_INT_COL in (1, 3)");
```
 in the benchmark setup to pollute the type profile and prevent the predicate evaluator from inlining demonstrates a larger difference, master here:

```
Benchmark               (_numRows)                                                                 (_query)  (_scenario)  Mode  Cnt   Score   Error  Units
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.001)  avgt    5  16.914 ± 2.387  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10     EXP(0.5)  avgt    5  40.642 ± 1.986  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.999)  avgt    5  35.026 ± 2.448  ms/op
```

Whereas with vectorisation (i.e. on this branch), throughput doesn't decrease because of the type pollution:

```
Benchmark               (_numRows)                                                                 (_query)  (_scenario)  Mode  Cnt   Score   Error  Units
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.001)  avgt    5  12.271 ± 1.857  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10     EXP(0.5)  avgt    5  30.771 ± 0.996  ms/op
BenchmarkQueries.query     1500000  SELECT SUM(RAW_INT_COL) FROM MyTable WHERE RAW_INT_COL BETWEEN 1 AND 10   EXP(0.999)  avgt    5  29.258 ± 3.856  ms/op
```

File: pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 1]
```diff
@@ -55,6 +58,33 @@ public SVScanDocIdIterator(PredicateEvaluator predicateEvaluator, ForwardIndexRe
...
+        limit = Math.min(_numDocs - _nextDocId, OPTIMAL_ITERATOR_BATCH_SIZE);
+        if (limit > 0) {
+          for (int i = 0; i < limit; i++) {
+            _batch[i] = _nextDocId + i;
+          }
+          batchSize = _valueMatcher.matchValues(limit, _batch);
+          _nextDocId += limit;
+          _numEntriesScanned += limit;
+        }
+      } while (limit > 0 & batchSize == 0);
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/SVScanDocIdIterator.java:74]: is this meant to be a bitwise AND ?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 2]
```diff
@@ -55,6 +58,33 @@ public SVScanDocIdIterator(PredicateEvaluator predicateEvaluator, ForwardIndexRe
...
+      _batchSize = batchSize;
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/SVScanDocIdIterator.java:75]: What would you call it instead?

**Review by siddharthteotia — APPROVED** [review]

**Comment by siddharthteotia** [comment]
> Also, whether saving a bit per value is worth the runtime cost of an extra dictionary reader implementation is an important design discussion.

I agree. Let's take this as a follow-up

**Pull request merged**